### ISSUE TITLE
Couldn't insert a record with nullable id

### ISSUE THREAD

**Issue by Deilan** [mbdavid/LiteDB#1052]
TL;DR: [Repository with the tests](https://github.com/deilan-issues-samples/LiteDB.Issues). More specifically:
- [LiteDatabaseInsertTests](https://github.com/deilan-issues-samples/LiteDB.Issues/blob/master/LiteDB.Issues.Tests/LiteDatabase/LiteDatabaseInsertTests.cs)
- [LiteRepositoryInsertTests](https://github.com/deilan-issues-samples/LiteDB.Issues/blob/master/LiteDB.Issues.Tests/LiteRepository/LiteRepositoryInsertTests.cs)

Model:
```csharp
public class Customer
{
    public int? Id { get; set; }
    public string Name { get; set; }
}
```

Test:
```csharp
var col = _liteDatabase.GetCollection<Customer>();

var customer = new Customer
{
    Name = "John Doe",
};

col.Insert(customer);

Assert.True(col.Exists(x => x.Id == customer.Id));
```

Actually `col.Insert` throws an exception:
> LiteDB.LiteException : Invalid BSON data type 'Null' on field '_id'.

Fails for both `LiteDatabase` and `LiteRepository`.

**Comment by stitch10925** [comment]
I don't really understand how this is an issue. It makes sense for LiteDB to not allow nullable Id's. Either you supply an Id which will be used, or you do not, in which case the Id will default to 0 and LiteDB will use auto-increment to create an Id for the entry.

**Comment by Deilan** [comment]
@stitch10925 There's no difference between `null` and `0` in that case: both are semantically equivalent. Also it's widely supported by other database access solutions. So common sense tells both cases should be equally supported.
P. S. Assume that you can't modify an existing entity, which already has a nullable Id. Having a dedicated entity and a specific mapping just for storage in LiteDB doesn't seem very reasonable.

**Comment by Deilan** [comment]
@mbdavid To create id automatically because it's null. Sorry for not being clear.

**Comment by mbdavid** [comment]
I agree with you: nullable ids must be auto id too. I fixing this, thanks

**Comment by padillah** [comment]
BTW: null and 0 are not the same. Null means it has not been set. 0 means it has been set... to 0.

I would think there's a strong argument for Auto Increment Id properties to be int? so we know they have not been set.

**Comment by falahati** [comment]
Hey @mbdavid,

I also strongly support what @padillah mentioned. `0` should be considered as a valid value for Id and rather `null` should be used as the AutoId default value. Had a problem with this in the past.

I also wanted to suggest adding support for `string` Ids and treat them as `guid` or `int` (Personally think `int` is better). In one of my projects, there is a base class for database access and it has a `string` property for Id for maximum flexibility, however, some of the child classes need to have AutoId. Currently `string`s are not considered as a valid type for AutoId.

I solved the problem by having a local clone of the project and implementing it, but still, I think it must be part of the library.

Thanks for the support and keeping this awesome project alive. It's been my number one choice in the majority of my projects since.

**Comment by mbdavid** [comment]
Hi @falahati, in v3 I used a custom function for custom type in conversion.

https://github.com/mbdavid/LiteDB/blob/v3/LiteDB/Mapper/BsonMapper.cs#L130

In this way is possible define how each data type will be increment if empty (and empty value are a defined by user)

But this occurs in "client" side (at database level) - not at engine level. In v4 I change this to add into engine level to support better concurrent. Now, in client side I just define when I consider an type as empty (like I defined `0` as empty in `int` type). Can see here:

https://github.com/mbdavid/LiteDB/blob/master/LiteDB/Database/Collections/Insert.cs#L87

In v4 (and v5) I just REMOVE `_id` from `BsonDocument` and pass to engine how new `_id` will be generate (as `BsonInt`, `BsonLong`, `BsonObjectId`, ...)

To accept `0` as a valid id, will need change this method. I will think about this and how this can be defined by used
